components/script/dom/bindings/str.rs

```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+    let mut is_match: bool = true;
+
+    for lang_range in lang_ranges {
+        // step 1
+        let css_tags: Vec<&str> = lang_range.split('\x2d').collect();
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if !(css_tags[0].eq_ignore_ascii_case(elem_tags[0]) || css_tags[0].eq_ignore_ascii_case("*")) {
```

Comment:
Let's not directly index the first element in the arrays. Instead, create an iterator on both of them, and use the following:
```rust
if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {
    // Do subtag matching...
}
```